src/main/scala/is/hail/distributedmatrix/RowMatrix.scala

```diff
@@ -105,4 +161,96 @@ class RowMatrix(val hc: HailContext,
       }.filter(_.nonEmpty)
     }.writeTable(path, hc.tmpDir, header, exportType)
   }
-}
\ No newline at end of file
+}
+
+// [`start`, `end`) is the row range of partition
+case class ReadBlocksAsRowsRDDPartition(index: Int, start: Long, end: Long) extends Partition
+
+class ReadBlocksAsRowsRDD(path: String,
+  sc: SparkContext,
+  partitionCounts: Array[Long],
+  gp: GridPartitioner) extends RDD[(Long, Array[Double])](sc, Nil) {
+  
+  private val partitionStarts = partitionCounts.scanLeft(0L)(_ + _)
+  
+  if (partitionStarts.last != gp.nRows)
+    fatal(s"Error reading BlockMatrix as RowMatrix: expected ${partitionStarts.last} rows in RowMatrix, but found ${gp.nRows} rows in BlockMatrix")
+
+  if (gp.nCols > Int.MaxValue)
+      fatal(s"Cannot read BlockMatrix with ${gp.nCols} > Int.MaxValue columns as a RowMatrix")
+  
+  private val nCols = gp.nCols.toInt
+  private val nBlockCols = gp.nBlockCols
+  private val blockSize = gp.blockSize
+
+  private val d = digitsNeeded(gp.numPartitions)
+  private val sHadoopBc = sc.broadcast(new SerializableHadoopConfiguration(sc.hadoopConfiguration))
+
+  protected def getPartitions: Array[Partition] = Array.tabulate(partitionStarts.length - 1)(pi => 
+    ReadBlocksAsRowsRDDPartition(pi, partitionStarts(pi), partitionStarts(pi + 1)))
+
+  def compute(split: Partition, context: TaskContext): Iterator[(Long, Array[Double])] = {
+    val ReadBlocksAsRowsRDDPartition(_, start, end) = split.asInstanceOf[ReadBlocksAsRowsRDDPartition]
+
+    var disArray = new Array[DataInputStream](nBlockCols)
```

Comment:
I don't like this name. Maybe `columnBlocks`?

Reply:
This sounds a bit weird with `blockCol` below. Maybe we can adopt the same index name conventions from the stuff I wrote for map & multiply?

Reply:
This is consistent with the current BlockMatrix and GridPartitioner

Reply:
but `disArray` is one of the least informative names possible. It's literally the type of the variable. `columnBlocks` is basically what this is.

Reply:
Ah, you're talking about `disArray`, not `nBlockCols`.